pubsub.go

```diff
@@ -359,29 +411,49 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
 	}
 	p.watchLk.Unlock()
 
-	sub, ok := p.subs[name]
+	ti, ok := p.topics[name]
 	if ok {
-		sub.Cancel()
-		delete(p.subs, name)
+		if err := p.closeTopic(name, ti); err != nil {
+			return false, err
+		}
+		<-ti.finished
 	}
 
-	if cancel, ok := p.cancels[name]; ok {
-		cancel()
-		delete(p.cancels, name)
+	return ok, nil
+}
+
+// closeTopic must be called under the PubSubValueStore's mutex
+func (p *PubsubValueStore) closeTopic(key string, ti *topicInfo) error {
+	ti.cancel()
+	ti.sub.Cancel()
+	ti.evts.Cancel()
+	if err := ti.topic.Close(); err != nil {
+		return err
 	}
 
-	return ok, nil
+	delete(p.topics, key)
+	return nil
 }
 
-func (p *PubsubValueStore) handleSubscription(ctx context.Context, sub *pubsub.Subscription, key string, cancel func()) {
-	defer sub.Cancel()
-	defer cancel()
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, ti *topicInfo, key string) {
+	defer func() {
+		ti.finished <- struct{}{}
+
+		p.mx.Lock()
+		defer p.mx.Unlock()
+
+		if err := p.closeTopic(key, ti); err != nil {
+			if err != context.Canceled && err != context.DeadlineExceeded {
```

Comment:
doing a type check here for the error doesn't feel great (what if pubsub decided to start wrapping errors, we'd have to pay attention). However, since there's nothing to do but log the messages this isn't too bad.

Unfortunately, since the pattern we use to deconstruct/close libp2p components is to have their parent context's expire and we generally have one cancellable parent context per application it's possible PubSub will get cancelled before the PubsubValueStore does.

Reply:
We can just check if the context has been canceled. If `ctx.Err() != nil`, we're shutting down and might as well ignore any errors.

Reply:
Wait, why is `closeTopic` returning errors _anyways_? Even when we're shutting down, `Close` should just close and go away.

Reply:
Ignoring it is probably fine.

Reply:
done through this file